modules/nlu/src/views/intents/slots/Slots.jsx

```diff
@@ -0,0 +1,220 @@
+  setSelection = (selectedText, intentEditor) => {
+    this.intentEditor = intentEditor
+
+    this.setState({ selectedText })
+  }
+
+  getSlots = () => {
+    return this.props.slots || []
+  }
+
+  updateSelectedSlot = (slot, index) => {
+    this.setState({
+      selectedSlot: slot,
+      selectedSlotIndex: index
+    })
+  }
+
+  changeSelectedSlot = step => {
+    const slots = this.getSlots()
+
+    let selectedSlotIndex = this.state.selectedSlotIndex + step
+    if (selectedSlotIndex >= slots.length) {
+      selectedSlotIndex = 0
+    } else if (selectedSlotIndex < 0) {
+      selectedSlotIndex = slots.length - 1
+    }
+
+    this.updateSelectedSlot(slots[selectedSlotIndex], selectedSlotIndex)
+  }
+
+  moveUp() {
+    this.changeSelectedSlot(-1)
+  }
+
+  moveDown() {
+    this.changeSelectedSlot(1)
+  }
+
+  onSlotSave = (slot, operation) => {
+    let slots = [...this.getSlots()]
+    let index = this.state.selectedSlotIndex
+    if (operation === 'modified') {
+      slots = [...slots.slice(0, this.state.selectedSlotIndex), slot, ...slots.slice(this.state.selectedSlotIndex + 1)]
+    } else {
+      index = slots.length
+      slots = [...slots, slot]
+    }
+
+    this.updateSelectedSlot(slot, index)
+    const oldname = this.selectedSlot ? this.selectedSlot.name : ''
+    this.props.onSlotsChanged && this.props.onSlotsChanged(slots, { operation, oldname, name: slot.name })
+  }
+
+  onSlotDeleted = slot => {
+    const slots = [...this.getSlots().filter(s => s.id !== slot.id)]
+    this.props.onSlotsChanged && this.props.onSlotsChanged(slots, { operation: 'deleted', name: slot.name })
+    this.updateSelectedSlot(slots.length ? slots[0] : null, slots.length ? 0 : null)
+  }
+
+  hasSelectedText = () => {
+    return this.state.selectedText && this.state.selectedText.length
+  }
+
+  // TODO on component will update, ==> update recommended action
+
+  renderWithSlots = () => {
+    const slots = this.getSlots()
+
+    if (this.hasSelectedText()) {
+      if (this.state.selectedText && this.state.selectedText.length) {
+        this.recommendedAction = () => {
+          this.tagSelectedText()
+          return 'create-entity'
+        }
+      }
+    }
+
+    return (
+      <div className={style.normalContainer}>
+        {this.hasSelectedText() && (
+          <Fragment>
+            <h4>
+              Tagging selected text <span className={style.selectionText}>"{this.state.selectedText}"</span>
+            </h4>
+            <hr />
+          </Fragment>
+        )}
+        <ul>
+          {slots.map((slot, i) => {
+            if (this.hasSelectedText()) {
+              return <ActionSlotItem key={slot.id} slot={slot} active={slot.id === this.state.selectedSlot.id} />
+            } else {
+              return (
+                <SlotItem
+                  key={slot.id}
+                  slot={slot}
+                  onDelete={this.onSlotDeleted}
+                  onEdit={this.showSlotModal.bind(this, slot, i)}
```

Comment:
You can use arrow functions for event handlers so you don't have to bind (when not performance-critical)